morepath/tests/test_directive.py

```diff
@@ -788,22 +818,61 @@ def a1_view(self, request):
         app.commit()
 
 
-def test_function_conflict():
-    class app(morepath.App):
+def test_function():
+    class App(morepath.App):
+        @reg.dispatch_method('a')
+        def func(self, a):
+            return "default"
+
+    class A(object):
         pass
 
+    @App.function(App.func, a=A)
+    def a_func(request):
```

Comment:
``@App.function`` is one of the places where we're breaking backward compatibility anyway.  I would take a step further and have it completely replaced it by ``@App.method``, or, equivalently, to ensure that decorated functions have ``app`` as their first argument:

```
@App.function(App.func, a=A)
    def a_func(app, request):
        ...
```